### PR TITLE
GH-2478 Handle conversion exception in AsyncRabbitTemplate

### PR DESCRIPTION
**Pull request by BenEfrati**
Previously, conversion errors in AsyncRabbitTemplate lead to AmqpReplyTimeoutException

Fixes https://github.com/spring-projects/spring-amqp/issues/2478

https://github.com/spring-projects/spring-amqp/issues/2478#issuecomment-2543745530

### CODE REVIEW

**Review by artembilan — CHANGES_REQUESTED** [review]
Please, run `./gradlew check` locally before pushing changes to PR.
Right now it even does not compile:
```
 /home/runner/work/spring-amqp/spring-amqp/spring-rabbit/src/main/java/org/springframework/amqp/rabbit/AsyncRabbitTemplate.java:614: error: cannot find symbol
						} catch (MessageConversionException e) {
						         ^
  symbol:   class MessageConversionException
  location: class AsyncRabbitTemplate
```

**Comment by BenEfrati** [comment]
Hi,
Sorry, at first I fork and created branch from old code, missed the import after sync with upstream.
https://github.com/BenEfrati/spring-amqp/commit/50452f89c3c306cc6dcad5b7dee27c036ffa81ac
Pushed the fixed code

File: spring-rabbit/src/test/java/org/springframework/amqp/rabbit/AsyncRabbitTemplateTests.java [hunk 1]
```diff
@@ -394,6 +396,29 @@ public void testStopCancelled() throws Exception {
 		assertThat(callback.result).isNull();
 	}
 
+	@Test
+	@DirtiesContext
```
Comment (artembilan) [inline spring-rabbit/src/test/java/org/springframework/amqp/rabbit/AsyncRabbitTemplateTests.java:400]: I wonder if we really need this `@DirtiesContext`.
I see that many other tests in this class have it, but that doesn't mean that it is right.
Would you mind to double check if it works without?
Just possible slight optimization in execution performance!

File: spring-rabbit/src/test/java/org/springframework/amqp/rabbit/AsyncRabbitTemplateTests.java [hunk 2]
```diff
@@ -394,6 +396,29 @@ public void testStopCancelled() throws Exception {
...
+		RabbitConverterFuture<String> replyFuture = this.asyncTemplate.convertSendAndReceive("conversionException");
+
+		final CountDownLatch cdl = new CountDownLatch(1);
+		final AtomicReference<Object> resultRef = new AtomicReference<>();
+        replyFuture.whenComplete((result, ex) -> {
+			resultRef.set(result);
+			cdl.countDown();
+		});
+        assertThat(cdl.await(10, TimeUnit.SECONDS)).isTrue();
+        assertThat(replyFuture).isCompletedExceptionally();
```
Comment (artembilan) [inline spring-rabbit/src/test/java/org/springframework/amqp/rabbit/AsyncRabbitTemplateTests.java:418]: May we also verify what exactly exception we got here?

**Review by artembilan — CHANGES_REQUESTED** [review]

File: spring-rabbit/src/test/java/org/springframework/amqp/rabbit/AsyncRabbitTemplateTests.java (hunk 1 above)
Comment (BenEfrati) [reply spring-rabbit/src/test/java/org/springframework/amqp/rabbit/AsyncRabbitTemplateTests.java:400]: Every test changes the `AsyncRabbitTemplate` state, so we must reset the context before each test. 
In my test, for example, I'm changing the messageConverter on the `RabbitTemplate`, so if my test will run first, all other tests will fail.

File: spring-rabbit/src/test/java/org/springframework/amqp/rabbit/AsyncRabbitTemplateTests.java (hunk 2 above)
Comment (BenEfrati) [inline spring-rabbit/src/test/java/org/springframework/amqp/rabbit/AsyncRabbitTemplateTests.java:418]:
```suggestion
		assertThat(replyFuture).failsWithin(Duration.ofSeconds(10))
				.withThrowableOfType(MessageConversionException.class)
				.withMessageContaining("Failed to convert message");
```
Something like this?

Comment (artembilan) [reply spring-rabbit/src/test/java/org/springframework/amqp/rabbit/AsyncRabbitTemplateTests.java:418]: Or never mind. I like it.
You have changed the logic, but at the same it is covered with that `failsWithin()` 😄 
Committing and merging.
Thank you!

File: spring-rabbit/src/test/java/org/springframework/amqp/rabbit/AsyncRabbitTemplateTests.java [hunk 3]
```diff
@@ -394,6 +396,29 @@ public void testStopCancelled() throws Exception {
...
+        assertThat(resultRef.get()).isNull();
```
Comment (artembilan) [inline spring-rabbit/src/test/java/org/springframework/amqp/rabbit/AsyncRabbitTemplateTests.java:419]: But this line has to be removed then 😄

Comment (BenEfrati) [reply spring-rabbit/src/test/java/org/springframework/amqp/rabbit/AsyncRabbitTemplateTests.java]: sure :) 

Do you want me to align other tests to use this `failsWithin`

```
void testReturn() {
...
		assertThat(future)
				.as("Expected exception")
				.failsWithin(Duration.ofSeconds(10))
				.withThrowableThat()
				.withCauseInstanceOf(AmqpMessageReturnedException.class)
				.extracting("routingKey")
				.isEqualTo(this.requests.getName() + "x");
```

instead 
```
try {
			future.get(10, TimeUnit.SECONDS);
			fail("Expected exception");
		}
		catch (ExecutionException e) {
			assertThat(e.getCause()).isInstanceOf(AmqpMessageReturnedException.class);
			assertThat(((AmqpMessageReturnedException) e.getCause()).getRoutingKey()).isEqualTo(this.requests.getName() + "x");
		}
```

Comment (artembilan) [reply spring-rabbit/src/test/java/org/springframework/amqp/rabbit/AsyncRabbitTemplateTests.java]: I would appreciate.
Thanks.
Those `try..catch` in tests is on my TODO list in favor of `assertThat()` when I touch these classes.
Although still would be with some code.
But now I have learned about `failsWithin()` which looks much cleaner. 😄

**Comment by artembilan** [comment]
@BenEfrati ,

thank you for contribution; looking forward for more!

**Pull request merged**